packages/web_framework/requirements.txt

```diff
@@ -1 +1,4 @@
+babel
 galaxy-data
+galaxy-web-stack
+paste
```

Comment:
I'll commit this - but I remain nervous given that we have like very specific requirements for SQLAlchemy for instance right (we pin it often)? We're duplicating that dependency over multiple applications instead of inheriting the single requirement from galaxy-data - which I sort of conceptually think of as bringing in our SQL Alchemy dependence.

Maybe we should find a longer term solution toward synchronizing Galaxy's Pipfile with the individual packages - maybe that would make me feel less yucky about this.

Do you have any ideas? I guess my two ideas would be to create Pipfiles for each individual project and then write a tool to combine them or to extend the Pipfile metadata with metadata about which Galaxy packages each dependency relies on. I had been leaning toward the former, but if you're wanting us to list each of these dependencies explicitly like this I'm guessing the latter would be a lot less duplication.

Reply:
My question would be: how will we actually be going to install Galaxy in the future?
1) as a collection of packages, in which case the pinning to exact versions will be done with pipenv (or poetry) on the top-level package
2) as a monolithic app (i.e. like now), in which case we would need to merge the individual package requirements into a Pipfile, and then pin to exact versions as usual. For the merge step, I suppose we could that with just `cat` and `sort -uf`

Reply:
That is a big question and I don't know that we can predict how everything will work in practice. I think the plan was that the ``galaxy`` package on PyPI would be published to by ``packages/meta/`` - which would depend on the individual packages (or really just galaxy-web-apps).

As like a starting point, I would imagine we could pin very specific packages for ``galaxy`` (i.e. our pinned requirements now) and keep everything else unpinned (except where they are needed). A second approach to that might be ``pip install galaxy[pinned]`` which would give you pinned versions of everything. I don't know if you can hide pinned versions behind a qualifier like that - but that would be nice IMO.

I'm not sure how we organize our dependencies for individual packages should depend a lot on those details though? I'm probably missing something though.

Reply:
> I'm not sure how we organize our dependencies for individual packages should depend a lot on those details though?

I agree it shouldn't (since we agree exact pins are appropriate only for the full application, not the packages). I asked the question simply because in case of option 1) we don't need to merge the various package requirements into a single Pipfile.

Reply:
> 1) we don't need to merge the various package requirements into a single Pipfile.

I see - I was assuming the git based install would be common ... for a long time after had packages available. And still available for a long time after that - I suspect then we'd still need to keep our scripts and such in place for managing dependencies.